tfhe/src/integer/server_key/radix_parallel/tests_signed/test_rotate.rs

```diff
@@ -0,0 +1,485 @@
+    };
+
+    if res < -modulus {
+        // rem_euclid(modulus) would also work
+        res = modulus + (res - -modulus);
+        overflowed = true;
+    } else if res > modulus - 1 {
+        res = -modulus + (res - modulus);
+        overflowed = true;
+    }
+    (res, overflowed)
+}
+
+pub(crate) fn signed_add_under_modulus(lhs: i64, rhs: i64, modulus: i64) -> i64 {
+    signed_overflowing_add_under_modulus(lhs, rhs, modulus).0
+}
+
+/// helper function to do a rotate left when the type used to store
+/// the value is bigger than the actual intended bit size
+pub(crate) fn rotate_left_helper(value: i64, n: u32, actual_bit_size: u32) -> i64 {
+    // We start with:
+    // [0000000000000|xxxx]
+    // 64           b    0
+    //
+    // rotated will be
+    // [0000000000xx|xx00]
+    // 64           b    0
+    let n = n % actual_bit_size;
+    let mask = 1i64.wrapping_shl(actual_bit_size) - 1;
+    let shifted_mask = mask.wrapping_shl(n) & !mask;
+
+    // Value maybe be negative and so, have its msb
+    // set to one, so use mask to only keep the part that interest
+    // us
+    let rotated = (value & mask).rotate_left(n);
+
+    let tmp = (rotated & mask) | ((rotated & shifted_mask) >> actual_bit_size);
+    // If the sign bit after rotation is one,
+    // then all bits above it needs to be one
+    let new_sign_bit = (tmp >> (actual_bit_size - 1)) & 1;
+    let mut pad = -new_sign_bit;
+    pad <<= actual_bit_size; // only bits above actual_bit_size should be set
+
+    pad | tmp
+}
+
+/// helper function to do a rotate right when the type used to store
+/// the value is bigger than the actual intended bit size
+pub(crate) fn rotate_right_helper(value: i64, n: u32, actual_bit_size: u32) -> i64 {
+    // We start with:
+    // [yyyyyyyyyyyy|xxxx]
+    // 64           b    0
+    // where xs are bits that we are interested in
+    // and ys are either 0 or 1 depending on if value is positive
+    //
+    // mask: [yyyyyyyyyyyy|mmmm]
+    // shifted_ mask: [mmyyyyyyyyyy|0000]
+    //
+    // rotated will be
+    // [xxyyyyyyyyyy|00xx]
+    // 64           b    0
+    //
+    // To get the 'cycled' bits where they should be,
+    // we get them using a mask then shift
+    let n = n % actual_bit_size;
+    let mask = 1i64.wrapping_shl(actual_bit_size) - 1;
+    // shifted mask only needs the bits that cycled
+    let shifted_mask = mask.rotate_right(n) & !mask;
+
+    // Value maybe be negative and so, have its msb
+    // set to one, so use mask to only keep the part that interest
+    // us
+    let rotated = (value & mask).rotate_right(n);
+
+    let tmp = (rotated & mask) | ((rotated & shifted_mask) >> (u64::BITS - actual_bit_size));
+    // If the sign bit after rotation is one,
+    // then all bits above it needs to be one
+    let new_sign_bit = (tmp >> (actual_bit_size - 1)) & 1;
+    let mut pad = -new_sign_bit;
+    pad <<= actual_bit_size; // only bits above actual_bit_size should be set
+
+    pad | tmp
+}
```

Comment:
Up till here you can just leave the code in tests_cases_signed and import the functions